services/director-v2/src/simcore_service_director_v2/modules/dask_client.py

```diff
@@ -0,0 +1,132 @@
+import logging
+from dataclasses import dataclass, field
+from typing import Callable, Dict, List, Union
+from uuid import uuid4
+
+from dask.distributed import Client, Future, fire_and_forget
+from fastapi import FastAPI
+from models_library.projects import ProjectID
+from simcore_service_director_v2.models.schemas.comp_scheduler import TaskIn
+from tenacity import before_sleep_log, retry, stop_after_attempt, wait_random
+
+from ..core.errors import ConfigurationError
+from ..core.settings import DaskSchedulerSettings
+from ..models.schemas.constants import UserID
+
+logger = logging.getLogger(__name__)
+
+
+dask_retry_policy = dict(
+    wait=wait_random(5, 8),
+    stop=stop_after_attempt(20),
+    before_sleep=before_sleep_log(logger, logging.WARNING),
+    reraise=True,
+)
+
+
+def setup(app: FastAPI, settings: DaskSchedulerSettings) -> None:
+    @retry(**dask_retry_policy)
+    def on_startup() -> None:
+        DaskClient.create(
+            app,
+            client=Client(
+                f"tcp://{settings.DASK_SCHEDULER_HOST}:{settings.DASK_SCHEDULER_PORT}",
+            ),
+            settings=settings,
+        )
+
+    async def on_shutdown() -> None:
+        app.state.dask_client.client.close()
+        del app.state.dask_client  # type: ignore
+
+    app.add_event_handler("startup", on_startup)
+    app.add_event_handler("shutdown", on_shutdown)
+
+
+@dataclass
+class DaskClient:
+    client: Client
+    settings: DaskSchedulerSettings
+
+    _taskid_to_future_map: Dict[str, Future] = field(default_factory=dict)
+
+    @classmethod
```

Comment:
I guess with the methods ``create`` and ``instance`` you aim to create a single instance of this client per app. 

IMO this pattern was nicely implemented in [AppDataMixin](https://github.com/ITISFoundation/osparc-simcore/blob/301d9d5d486e11f3e6353debb76f6f02dc07ca12/services/api-server/src/simcore_service_api_server/utils/app_data.py#L7)

See usage in [test_utils_app.data.py](https://github.com/ITISFoundation/osparc-simcore/blob/301d9d5d486e11f3e6353debb76f6f02dc07ca12/services/api-server/tests/unit/test_utils_app_data.py)

Perhaps we could move this to ``servicelib``?

Reply:
agree. but for later